### ISSUE TITLE
Factory does not create Password Element if only Password Attribute is present

### ISSUE THREAD

**Issue by claytondaley** [zendframework/zf2#7166]
Issue #2602 added desirable security functionality to the Password Element including:

 - If an Edit form is built from a database object, the password will automatically be excluded from the data sent to the browser

It was later noted that the password attribute (frequently used to identify a password field) is not considered by the Factory:

    array(
        'name' => 'something',
        'attributes' => array(
            'type' => 'password',
        ),
        // ...
    )

This results in a `\Zend\Form\Element` instead of a `\Zend\Form\Element\Password`.  To get a password element, the user must explicitly identify the password type in the parent array:

    array(
        'name' => 'something',
        'type' => 'Password',
        // ...
    )

To be fair, the distinction is real and could theoretically be a desired behavior.  

 - At minimum, I would recommend adding a special warning in the documentation, possibly under [Factory-Backed Form Extension](http://framework.zend.com/manual/current/en/modules/zend.form.quick-start.html#factory-backed-form-extension).
 - However, I also believe this is an obvious spot to "nudge" users by making the most secure assumption (password attribute intends to create a password element).  Obviously, a user could override this setting by explicitly asking for a `type => text`.

**Comment by Martin-P** [comment]
> It was later noted that the password attribute (frequently used to identify a password field) is not considered by the Factory

Works fine for me when I use only a type attribute:
```php
$form = new \Zend\Form\Form();
$form->add(array(
    'name' => 'foo',
    'attributes' => array(
        'type' => 'password',
    ),
));

var_dump($this->formElement($form->get('foo')));
```
Output:
```html
<input name="foo" type="password" value="">
```
Do you use and third party modules? If so, deactivate them to make sure that is not causing the problem.

**Comment by claytondaley** [comment]
You get the attribute, but the underlying Element is wrong so you don't get the full Password Element behaviors.  For example, if you attach this element to a database object, it will actually fill the field (at the browser) with the (usually hashed) password stored in the database.  As indicated in #2602, the actual Password Element knows not to do this.

**Comment by Martin-P** [comment]
I see your point now and can confirm the bug:
```php
<?php
$form = new \Zend\Form\Form();
$form->add(array(
    'name' => 'foo',
    'attributes' => array(
        'type'  => 'password',
        'value' => 'secret',
    ),
));

var_dump($this->formElement($form->get('foo')));
```
Output:
```html
<input name="foo" type="password" value="secret">
```
I reported the original issue #2602 but never noticed it was not completely fixed.

Perhaps it is better to change the view helper itself instead of the form element. That way it is fixed at one location instead of two separate locations. In the test I also see the password is reset *only* if the form is invalid: [`ZendTest\Form\FormTest::testResetPasswordValueIfFormIsNotValid()`](https://github.com/zendframework/zf2/blob/f5f5cee84c633161e0b659b51777f48baf1ab94b/tests/ZendTest/Form/FormTest.php#L1107-L1128). That makes no sense to me. I think the value of a password field should *never* be rendered regardless whether the form is valid or not.

**Comment by claytondaley** [comment]
I'm glad you were able to reproduce so quickly as I'm definitely not familiar enough with the Form internals to have done so.  Any solution is great so long as treating it as another valid type like an Element (i.e. calling the Element view helper on it) doesn't bypass the fix.

**Comment by Martin-P** [comment]
Fixed in #7171.
